links/link_veth.go

```diff
@@ -109,70 +107,145 @@ func (*LinkVEth) GetType() LinkType {
+	if peerEp.AutoDeployWithAEnd() {
+		return l.deployBEnd(ctx, peerEp)
 	}
 
-	// both ends of the link need to be moved to the relevant network namespace
+	return nil
+}
+
+func (l *LinkVEth) deployBEnd(ctx context.Context, ep Endpoint) error {
+
+	idx, err := l.getEndpointIndex(ep)
+	if err != nil {
+		return err
+	}
+
+	peerEp := l.Endpoints[(idx+1)%2]
+
+	log.Infof("Assigning Endpoint: %s ( --> %s )", ep, peerEp)
+
+	// retrieve the netlink.Link for the provided Endpoint
+	link, err := utils.LinkByNameOrAlias(ep.GetRandIfaceName())
+	if err != nil {
+		return err
+	}
+
+	// disable TXOffloading
+	if err := utils.EthtoolTXOff(ep.GetRandIfaceName()); err != nil {
+		return err
+	}
+
+	// the link needs to be moved to the relevant network namespace
 	// and enabled (up). This is done via linkSetupFunc.
 	// based on the endpoint type the link setup function is different.
 	// linkSetupFunc is executed in a netns of a node.
-	for idx, link := range []netlink.Link{linkA, linkB} {
-		// if the node is a regular namespace node
-		// add link to node, rename, set mac and Up
-		err = l.Endpoints[idx].GetNode().AddLinkToContainer(ctx, link,
-			SetNameMACAndUpInterface(link, l.Endpoints[idx]))
-		if err != nil {
-			return err
+	// if the node is a regular namespace node
+	// add link to node, rename, set mac and Up
+	err = ep.GetNode().AddLinkToContainer(ctx, link,
+		SetNameMACAndUpInterface(link, ep))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getEndpointIndex returns the index of the ep endpoint belonging to l link.
+// An error is returned when the ep is not part of the l's endpoints.
+func (l *LinkVEth) getEndpointIndex(ep Endpoint) (int, error) {
+	for idx, e := range l.Endpoints {
+		if e == ep {
+			return idx, nil
 		}
 	}
 
-	l.DeploymentState = LinkDeploymentStateDeployed
+	// if the endpoint is not part of the link
+	// build a string list of endpoints and return a meaningful error
+	epStrings := []string{}
+	for _, e := range l.Endpoints {
+		epStrings = append(epStrings, e.String())
+	}
 
-	return nil
+	return -1, fmt.Errorf("endpoint %s does not belong to link [ %s ]", ep.String(), strings.Join(epStrings, ", "))
+
+}
+
+// Deploy deploys the veth link by creating the A and B sides of the veth pair independently
+// based on the calling endpoint.
+func (l *LinkVEth) Deploy(ctx context.Context, ep Endpoint) error {
+	// since each node calls deploy on its links, we need to make sure that we only deploy
+	// the link once, even if multiple nodes call deploy on the same link.
+	l.deployMutex.Lock()
+	defer l.deployMutex.Unlock()
+
+	// first we need to check that the provided ep is part of this link
+	_, err := l.getEndpointIndex(ep)
+	if err != nil {
+		return err
+	}
+
+	// The first node to trigger the link creation will call deployAEnd,
+	// subsequent (the second) call will end up in deployBEnd.
+	switch l.DeploymentState {
+	case LinkDeploymentStateDeployed:
```

Comment:
this variable name seems misleading. At this point in time we haven't deployed the link, but just the A end of it.

Feels like this should be `LinkDeploymentStateAEndDeployed`
then once the B End is deployd the link reaches `LinkDeploymentStateDeployed`?

Reply:
Yes you are right... the state was there as is and I lazily missused it.

Reply:
Done